### ISSUE TITLE
Support PostgreSQL using ProvidePasswordCallback for azure ad

### ISSUE THREAD

**Issue by mikara89** [dotnetcore/CAP#1310]
I am using option.UseEntityFramework(...) with Postgres and 

```
 services.AddCap(options =>
            {
                options.UseEntityFramework<CapDbContext>(opt => opt.Schema = "Cap");
                options.UseInMemoryMessageQueue();
                options.UseDashboard(options =>
                {
                    options.UseAuth = true;
                    options.DefaultAuthenticationScheme = "...";
                });

            });
```
```
private static IServiceCollection AddPGSQL<T>(this IServiceCollection services, string connectionString) 
            where T : DbContext
        {
            services.AddDbContextFactory<T>(m => m.UseNpgsql(connectionString, e => {
                    e.ProvidePasswordCallback((h, p, d, u) => {
                        var tokenCredential = new DefaultAzureCredential();
                        var AccessToken = tokenCredential.GetToken(new Azure.Core.TokenRequestContext(
                            new[] { "https://ossrdbms-aad.database.windows.net/.default" }
                            ));

                        return AccessToken.Token;
                    });
                    e.MigrationsAssembly(typeof(T).Assembly.FullName);
                }), ServiceLifetime.Scoped);
            return services;
        }
```
for using access token as password on db but cap is not triggering `ProvidePasswordCallback` at all.
Is this bug or what?

**Comment by yang-xiaodong** [comment]
I was not aware of ProvidePasswordCallback before. We obtain the connection string from DbContext for ADO.NET database connections, and we do not dynamically update the connection string during its use.
https://github.com/dotnetcore/CAP/blob/191937eb799e7c847abfd7b36fbfdad55a2cf8d1/src/DotNetCore.CAP.PostgreSql/CAP.PostgreSqlOptions.cs#L42

If you are interested add this feature, welcome submit a PR.

**Comment by mikara89** [comment]
> I was not aware of ProvidePasswordCallback before. We obtain the connection string from DbContext for ADO.NET database connections, and we do not dynamically update the connection string during its use.
> 
> https://github.com/dotnetcore/CAP/blob/191937eb799e7c847abfd7b36fbfdad55a2cf8d1/src/DotNetCore.CAP.PostgreSql/CAP.PostgreSqlOptions.cs#L42
> 
> If you are interested add this feature, welcome submit a PR.

Thanks on replay. I have find out not long ago for use case where security is not allowing static password, and only solution is using Manage Identity. You can see that password is access token. When I have time I will provide PR for covering this use case. MSSQL and MySQL are using "..;Authentication=Active Directory Interactive;" as part of connection string to use Manage Identity but that is not use case for PostgreSQL.

**Comment by jonekdahl** [comment]
I'm running into this when attempting to use CAP with EFCore, PostgreSQL and Azure managed identity, so I would be interested in submitting a PR to support this use case. In my case I'm planning to leverage [NpgsqlDataSourceBuilder.UsePeriodicPasswordProvider(...)](https://www.npgsql.org/doc/api/Npgsql.NpgsqlDataSourceBuilder.html#Npgsql_NpgsqlDataSourceBuilder_UsePeriodicPasswordProvider_System_Nullable_Func_Npgsql_NpgsqlConnectionStringBuilder_CancellationToken_ValueTask_System_String____TimeSpan_TimeSpan_) to keep the password up to date.

@yang-xiaodong Do you have a solution in mind to support this use case? If not, I can start digging around to see what the options are.

At first glance the `DatabaseFacade` from where the connection string is currently fetched does not provide access to the password provider that was configured on the data source.

**Comment by yang-xiaodong** [comment]
@jonekdahl  That's great, PR is welcome

**Comment by jonekdahl** [comment]
@yang-xiaodong Alright,.I'll investigate the options. Should I open a new issue or can this one be reopened?

**Comment by jonekdahl** [comment]
I did a proof of concept on the data source idea [here](https://github.com/dotnetcore/CAP/compare/master...jonekdahl:CAP:npgsql-data-source-support). If this seems like a good way forward I could open a PR and we can take it from there.

One thing I still haven't figured out is whether it's possible to support the `UseEntityFramework<'dbcontext>()` way of configuring CAP when EFCore is configured using a data source.

**Comment by yang-xiaodong** [comment]
@jonekdahl Using the `UseEntityFramework<T>` configuration method compared to using UsePostgresSql is only convenient for retrieving the ConnectionString and serves no other purpose. The code is here: https://github.com/dotnetcore/CAP/blob/2194ef53f738f50a1167f301d334d23170fc1c4c/src/DotNetCore.CAP.PostgreSql/CAP.PostgreSqlOptions.cs#L30-L43. 

I've looked into the support method in EF Core, and to support `ProvideClientCertificatesCallback` and `RemoteCertificateValidationCallback`, a simple way is to depend on the NpgsqlDbContextOptionsBuilder configuration option from the `Npgsql.EntityFrameworkCore.PostgreSQL` package, but I think this might not be a good approach. Therefore, I believe we can choose not to offer this feature in the EF configuration. If users want to use dynamic passwords, they can just use the `UsePostgreSql` configuration method. What do you think?

**Comment by jonekdahl** [comment]
@yang-xiaodong I agree, being able to use `UseEntityFramework<T>` in the common case of the password being part of the connection string is certainly convenient but having to drop down to something like `UsePostgreSql(opts => options.DataSource = ...)` for the more complex cases is pretty reasonable.

I'll open a PR then.

**Comment by yang-xiaodong** [comment]
Fixed in version 8.1.0
